common/src/main/scala/uk/ac/wellcome/dynamo/VersionedDao.scala

```diff
@@ -2,57 +2,164 @@ package uk.ac.wellcome.dynamo
 
 import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
 import com.google.inject.Inject
+import com.gu.scanamo.query.{KeyEquals, UniqueKey}
 import com.gu.scanamo.syntax.{attributeExists, not, _}
+import com.gu.scanamo.update.UpdateExpression
 import com.gu.scanamo.{DynamoFormat, Scanamo, Table}
 import com.twitter.inject.Logging
+import shapeless.{HList, LabelledGeneric, Poly1, Witness}
+import shapeless.labelled.FieldType
 import uk.ac.wellcome.models._
+import shapeless._
+import shapeless.ops.hlist.{LeftFolder, Mapper}
+import shapeless.ops.record.Remover
+import shapeless.ops.tuple.ToTraversable
+import shapeless.record._
+import uk.ac.wellcome.dynamo.UpdateExpressionGenerator.toUpdateExpressions
 import uk.ac.wellcome.models.aws.DynamoConfig
+import uk.ac.wellcome.type_classes.IdGetter.{createIdGetter, IdKey}
+import uk.ac.wellcome.type_classes.{IdGetter, VersionGetter, VersionUpdater}
 import uk.ac.wellcome.utils.GlobalExecutionContext.context
 
+import scala.collection.generic.CanBuildFrom
 import scala.concurrent.Future
 
+trait UpdateExpressionGenerator[T] {
+  def generateUpdateExpression(t: T): Option[UpdateExpression]
+}
+
+object UpdateExpressionGenerator {
```

Comment:
This is what creates the update expressions that we can throw the type checker at, yes?

So the theory roughly goes: the "Witness" allows us to find objects who have an attribute "id" of any type, and then … waves hands … an update expression pops out?